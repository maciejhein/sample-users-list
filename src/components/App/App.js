import React from "react";

import { useUsers } from "../../hooks";
import Users from "../Users";

const App = () => {
  const { users, error } = useUsers();

  const renderAppContent = () => {
    if (!users && !error) {
      return <div>Loading...</div>;
    }

    if (error) {
      throw error;
    }

    return <Users users={users} />;
  };

  return <main className="main-container">{renderAppContent()}</main>;
};

export default App;
