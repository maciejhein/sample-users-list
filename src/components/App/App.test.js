import React from "react";
import { render } from "@testing-library/react";

import App from "./App";

const mockedArticleValue = jest.fn();
const mockedErrorValue = jest.fn();

jest.mock("../Users", () => () => "Users");

jest.mock("../../hooks", () => ({
  useUsers: () => ({
    users: mockedArticleValue(),
    error: mockedErrorValue(),
  }),
}));

describe("COMPONENT - App", () => {
  it("renders App component with Users", () => {
    mockedArticleValue.mockImplementation(() => "fooUsers");
    mockedErrorValue.mockImplementation(() => null);

    const { asFragment } = render(<App />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders App component with loading", () => {
    mockedArticleValue.mockImplementation(() => null);
    mockedErrorValue.mockImplementation(() => null);

    const { asFragment } = render(<App />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("throws an Error if useUsers failed", () => {
    let consoleErrorSpy;

    consoleErrorSpy = jest.spyOn(console, "error").mockImplementation(() => {
      // Allows suppressing error logs
    });

    mockedArticleValue.mockImplementation(() => null);
    mockedErrorValue.mockImplementation(() => new Error("Foo error"));

    expect(() => render(<App />)).toThrow();

    consoleErrorSpy.mockRestore();
  });
});
