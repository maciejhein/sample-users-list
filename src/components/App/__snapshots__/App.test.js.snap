// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COMPONENT - App renders App component with Users 1`] = `
<DocumentFragment>
  <main
    class="main-container"
  >
    Users
  </main>
</DocumentFragment>
`;

exports[`COMPONENT - App renders App component with loading 1`] = `
<DocumentFragment>
  <main
    class="main-container"
  >
    <div>
      Loading...
    </div>
  </main>
</DocumentFragment>
`;
